src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
             TimeSpan timeout,
             bool disableAia)
         {
-            throw new NotImplementedException(nameof(BuildChain));
+            var chainPal = new AndroidCertPath();
+            try
+            {
+                chainPal.Initialize(cert, extraStore, customTrustStore, trustMode);
+                chainPal.Evaluate(verificationTime, applicationPolicy, certificatePolicy, revocationMode, revocationFlag);
+            }
+            catch
+            {
+                chainPal.Dispose();
+                throw;
+            }
+
+            return chainPal;
+        }
+
+        private sealed class AndroidCertPath : IChainPal
+        {
+            public X509ChainElement[]? ChainElements { get; private set; }
+            public X509ChainStatus[]? ChainStatus { get; private set; }
+
+            public SafeX509ChainHandle? SafeHandle => null;
+
+            private SafeX509ChainContextHandle? _chainContext;
+            private bool _isValid;
+
+            public void Dispose()
+            {
+                if (_chainContext != null)
+                {
+                    _chainContext.Dispose();
+                }
+            }
+
+            public bool? Verify(X509VerificationFlags flags, out Exception? exception)
+            {
+                Debug.Assert(_chainContext != null);
+                exception = null;
+
+                if (!_isValid)
+                {
+                    // There is no way to bypass time, trusted root, name, or policy constraint
+                    // validation on Android. It will not build any chain without these all
+                    // being valid.
+                    X509VerificationFlags[] unsupportedFlags = new X509VerificationFlags[]
+                    {
+                        X509VerificationFlags.IgnoreNotTimeValid,
+                        X509VerificationFlags.AllowUnknownCertificateAuthority,
+                        X509VerificationFlags.IgnoreInvalidName,
+                        X509VerificationFlags.IgnoreInvalidPolicy,
+                    };
+                    foreach (X509VerificationFlags unsupported in unsupportedFlags)
+                    {
+                        if ((flags & unsupported) == unsupported)
+                        {
+                            exception = new PlatformNotSupportedException(SR.Format(SR.Chain_VerificationFlagNotSupported, unsupported));
+                            return default(bool?);
+                        }
+                    }
+
+                    return false;
+                }
+
+                return ChainVerifier.Verify(ChainElements!, flags);
+            }
+
+            internal void Initialize(
+                ICertificatePal cert,
+                X509Certificate2Collection? extraStore,
+                X509Certificate2Collection customTrustStore,
+                X509ChainTrustMode trustMode)
+            {
+                List<IntPtr> extraCerts = new List<IntPtr>() { cert.Handle };
```

Comment:
There's a bit of a race condition problem here with disposing.  A more robust solution would be to have a List<SafeX509Handle> that holds all of the Pal SafeHandle values, they get DangerousAddRef'd in the beginning, then DangerousRelease'd in a finally.  (Once they're all add-ref'd then make an array for the IntPtr values).

Alternatively, the P/Invoke marshaller might inherently support `SafeHandle[]` as doing the upref/downref thing and marshalling as the handle values.  (I don't remember if it does or not).